---
title: "DSP_CNV_e1"
author: "Zack Lewis"
date: "11/24/2020"

header-includes:
   - \usepackage[default]{sourcesanspro}
   - \usepackage[T1]{fontenc}
mainfont: SourceSansPro
output:
  pdf_document:
    extra_dependencies: ["flafter"]
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(echo = TRUE, dev="cairo_pdf")
```

## Read in data

```{r warning=FALSE,echo=FALSE, include=FALSE}
library(tidyverse)
library(dplyr)
#library(egg) # for ggarrange
library(pheatmap)
library(psych) # for geometric.mean
library(reshape2)
library(RColorBrewer)
library(colorRamps)
library(qualpalr)
library(kableExtra)
library(ggpubr) # ggarrange and for common labels on ggarrrange plots

```

## Experimental conditions

- Four slides at different probe concentrations
- One 16 nM slide was denatured at 92C instead of 78C (Slide 3)


```{r expt conditions}
slide_number <- c(1,2,3,5)
denaturation_temperature <- c(78,78,92,78)
probe_concentration <-c(8,16,16,32)

tibble(slide_number,denaturation_temperature,probe_concentration)
```

```{r warning=FALSE}
counts <- read_tsv("DSP_CNV_e1/output_dedup.txt")
counts_long <- pivot_longer(counts, cols=-Sample_ID) %>% dplyr::rename(rts_id = name, count = value)

aoi <- read_csv("AOI_descriptions.csv")
#rts <- read_csv("rts_assignment.csv") %>% dplyr::rename(rts_id = 'RTS ID')
rts<- read_csv("rts.csv")

counts_described <- left_join(aoi,counts_long)

counts_assigned <- left_join(counts_described,rts)

#write_csv(counts_assigned, path = "DSP_CNV_exp1_counts.csv")

```

```{r fig.cap="ERBB2 raw counts."}
ggplot(counts_assigned %>% 
         filter(geneName %in% "ERBB2") %>% 
         filter(Size == 600),
       aes(x = AOI, y = count, col= Tissue, shape = as.factor(Slide))) + 
  geom_point() +
  labs(title="ERBB2 raw counts")
```

```{r fig.cap="CDKN2A raw counts"}
ggplot(counts_assigned %>% 
         filter(geneName %in% "CDKN2A") %>% 
         filter(Size == 600),
       aes(x = AOI, y = count, col= Tissue, shape = as.factor(Slide))) + 
  geom_point() +
  labs(title="CDKN2A raw counts")
```

```{r fig.cap="gap0121_347804.1:2 raw counts"}
ggplot(counts_assigned %>% 
         filter(geneName %in% "gap0121_347804.1:2") %>% 
         filter(Size == 600),
       aes(x = AOI, y = count, col= Tissue, shape = as.factor(Slide))) + 
  geom_point() +
  labs(title="invariant gap0121_347804.1:2 (chr8) raw counts")
```

### Make subsets for each slide

```{r}
# slide 1
S1 <- counts_assigned %>% filter(Slide == 1)
# slide 2
S2 <- counts_assigned %>% filter(Slide == 2)
# slide 3
S3 <- counts_assigned %>% filter(Slide == 3)
# slide 5
S5 <- counts_assigned %>% filter(Slide == 5)
```


## Multipanel plots

### Slide 1 - 8 nM

```{r fig.cap="Normalized counts for slide 1 (8 nM). Counts are divided by the geometric mean of the ERCC negative counts for that slide.", fig.width = 11, fig.height=7}

# find the geometric mean of the negative counts
ercc_geomean_S1 <- geometric.mean(
  S1 %>% 
    filter(category %in% "ercc") %>% 
    filter(Size==600) %>% 
    select(count) %>% as.matrix()
  )

## find the sd
ercc_sd_S1 <- sd(
    S1 %>% 
      filter(category %in% "ercc") %>% 
      filter(Size==600) %>% 
      select(count) %>% as.matrix()
    )


variantPlot_S1 <- S1 %>% 
  filter(category %in% c("variant")) %>% 
  filter(Size == 600) %>% 
  ggplot(aes(x = geneName, y =log2( count/ercc_geomean_S1), col = Tissue)) + geom_jitter() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(limits = c(0,5))+
  labs(title="Slide 1 - variant") 


multicopyPlot_S1 <- S1 %>% 
  filter(category %in% c("multicopy")) %>% 
  filter(Size == 600) %>% 
  ggplot(aes(x = geneName, y = log2(count/ercc_geomean_S1), col = Tissue)) + geom_jitter() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(limits = c(0,5)) +
  labs(title="multicopy")

invariantPlot_S1 <- S1 %>% 
  filter(category %in% "invariant") %>% 
  filter(Size == 600) %>% 
  ggplot(aes(x = chr, y = log2(count/ercc_geomean_S1), col = Tissue)) + geom_jitter() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(limits = c(0,5)) +
  labs(title="invariant") 


slide1<- ggarrange(
  variantPlot_S1, invariantPlot_S1, multicopyPlot_S1,
  nrow = 1, common.legened = TRUE
  )


#ggsave(slide1,filename = "slide_1_variants_invariants_transformed.png",device="png",width = 14, height=6)

```

### Slide 5 - 32 nM

```{r fig.cap="Normalized counts for slide 5 (32 nM). Counts are divided by the geometric mean of the ERCC negative counts for that slide.", fig.width = 11, fig.height=7}

# multi panel plot for slide 5
neg_mean_S5 <- geometric.mean(
  S5 %>% 
    filter(category %in% "ercc") %>% 
    filter(Size==600) %>% 
    select(count) %>% as.matrix()
  )


variantPlot_S5 <- S5 %>% 
  filter(category %in% c("variant")) %>% 
  filter(Size == 600) %>% 
  ggplot(aes(x = geneName, y = count/neg_mean_S5, col = Tissue)) + geom_jitter() +
  theme(legend.position = "none",axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(limits = c(0,6))+
  labs(title="Slide 5 - variant") 

multicopyPlot_S5 <- S5 %>% 
  filter(category %in% c("multicopy")) %>% 
  filter(Size == 600) %>% 
  ggplot(aes(x = geneName, y = count/neg_mean_S5, col = Tissue)) + geom_jitter() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(limits = c(0,6)) +
  labs(title="multicopy")

invariantPlot_S5 <- S5 %>% 
  filter(category %in% "invariant") %>% 
  filter(Size == 600) %>% 
  ggplot(aes(x = chr, y = count/neg_mean_S5, col = Tissue)) + geom_jitter() +
  theme(legend.position = "none",axis.text.x = element_text(angle = 90)) +
  scale_y_continuous(limits = c(0,6)) +
  labs(title="invariant")


slide5<- ggarrange(
  variantPlot_S5,invariantPlot_S5,multicopyPlot_S5, 
  nrow = 1
  )

#ggsave(slide5,filename = "slide_5_variants_invariants.png",device="png",width = 10, height=6)
```

## Plotting the limit of detection

- Black line is the geometric mean of the ercc controls for that slide
- Blue dashed line is plus two standard deviations

```{r}
# function to plot lines for geomean ercc and +2 SD

LOD_plot <- function(dataset, category_to_plot){
  
  # calculate the ERCC mean and SD
  ## find the geometric mean of the negative counts
    ercc_geomean <- geometric.mean(
      dataset %>% 
        filter(category %in% "ercc") %>% 
        filter(Size==600) %>% 
        select(count) %>% as.matrix()
      )
  ## find the sd
    ercc_sd <- sd(
      dataset %>% 
        filter(category %in% "ercc") %>% 
        filter(Size==600) %>% 
        select(count) %>% as.matrix()
      )
  
  # generate the plot
  dataset %>% 
    filter(category %in% category_to_plot) %>% 
    filter(Size == 600) %>% 
    ggplot(aes(x = geneName, y = log2(count), col = Tissue)) + geom_jitter(size = 1) +
    theme(axis.text.x = element_text(angle = 90)) +
    labs(title=paste(category_to_plot)) +
    geom_hline(yintercept=log2(ercc_geomean)) +
    geom_hline(yintercept = log2(ercc_geomean + 2*ercc_sd), color = "blue",lty = 2)
}

```



```{r fig.width = 11, fig.height=11, fig.cap="Slide 1. Log-transformed counts for each category with black lines designating the geometric mean of the negative control counts and the blue dashed line showing 2 standard deviations from the geometric mean."}
# slide 1
S1 <- counts_assigned %>% filter(Slide == 1)
ggarrange(
LOD_plot(S1, "invariant")+theme(axis.text.x = element_blank()),
LOD_plot(S1, "variant"),
LOD_plot(S1, "ercc")+theme(axis.text.x = element_blank())) %>% ggsave(filename = "LOD_Plot_S1.png",device="png",height=10,width=7)
```

```{r fig.width = 11, fig.height=11, fig.cap="Slide 1 with SKBR3 and MCF7 removed because of their high and low counts, respectively. Log-transformed counts for each category with black lines designating the geometric mean of the negative control counts and the blue dashed line showing 2 standard deviations from the geometric mean."}


# SKBR3 has pretty high ercc counts across the board. 
# MCF7 is low across the board
## by eliminating both, we should lower the SD

S1_noSKBR3 <- counts_assigned %>% filter(Slide == 1) %>% filter(Tissue != "SKBR3") %>%
   filter(Tissue != "MCF7") # %>%
  # filter(Tissue != "CD45H") %>%
  # filter(Tissue != "CD45L")

S1_skbr3_rem <- ggarrange(
  LOD_plot(S1_noSKBR3, "invariant")+ 
    theme(axis.text.x = element_blank()) + 
    scale_y_continuous(limits = c(9,13)),
  LOD_plot(S1_noSKBR3, "variant")+ 
    scale_y_continuous(limits = c(9,13)),
  LOD_plot(S1_noSKBR3, "ercc") + 
    theme(axis.text.x = element_blank()) + 
    scale_y_continuous(limits = c(9,13)))

ggsave(S1_skbr3_rem, filename = "LOD_plot_S1_removedSKBR3_MCF7_BC.png",device ="png", 
       height = 10, width = 7)
```

```{r fig.width = 11, fig.height=11, fig.cap="Slide 2. Log-transformed counts for each category with black lines designating the geometric mean of the negative control counts and the blue dashed line showing 2 standard deviations from the geometric mean."}

# slide 2
S2 <- counts_assigned %>% filter(Slide == 2)
ggarrange(
LOD_plot(S2, "invariant")+theme(axis.text.x = element_blank()),
LOD_plot(S2, "variant"))
```

```{r fig.width = 11, fig.height=11, fig.cap="Slide 3. Log-transformed counts for each category with black lines designating the geometric mean of the negative control counts and the blue dashed line showing 2 standard deviations from the geometric mean."}

# slide 3
ggarrange(
LOD_plot(S3, "invariant")+theme(axis.text.x = element_blank()),
LOD_plot(S3, "variant"))
```

```{r fig.width = 11, fig.height=11, fig.cap="Slide 5. Log-transformed counts for each category with black lines designating the geometric mean of the negative control counts and the blue dashed line showing 2 standard deviations from the geometric mean."}

#slide 5
ggarrange(
LOD_plot(S5, "invariant")+theme(axis.text.x = element_blank()),
LOD_plot(S5, "variant"))

```


## Normalization by chromosome count


#### Chromosome 17 probes:

```{r fig.cap="Chromosome 17 probe counts normalized to the counts for invariant probes on that chromosome."}

# get the mean count for chromosome 17 on Slide 1

chr17_S1_count <- S1 %>% filter(category %in% "invariant") %>%
  filter(Size == 600) %>%
  filter(chr %in% "chr17") %>% 
  select(count) %>%
  as.matrix() %>%
  geometric.mean()
chr17_S1_count

chr17_ercc_count <- S1 %>% filter(category %in% "ercc") %>%
  filter(Size == 600) %>%
  select(count) %>%
  as.matrix() %>%
  geometric.mean()  
chr17_ercc_count

# plot counts corrected by the mean chromosome 17 count

chr17_probes <- S1 %>%
  filter(chr %in% "chr17") %>%
  filter(Size == 600) %>%
  ggplot(aes(x = geneName, y = log2(count/chr17_S1_count), col = Tissue)) + geom_jitter(size = 2) +
  theme(axis.text.x = element_text(angle = 90)) +
#  scale_y_continuous(limits = c(0,6))+
  labs(title="chromosome 17")
chr17_probes 
#chr17_probes %>% ggsave(filename = "chromosome17_probes.png",device="png")

```

- Affymetrix data are normalized by `log2(CN/2))`
- COSMIC data use CN counts: 

    Minor Allele: the number of copies of the least frequent allele eg if ABB, minor allele = A ( 1 copy) and major allele = B ( 2 copies)
    Copy Number: the sum of the major and minor allele counts eg if ABB, copy number = 3

#### Chromosome 9 probes: CDKN2A should have low counts

```{r fig.cap="Chromosome 9 probe counts normalized to the counts for invariant probes on that chromosome." }
# CDKN2A is on Chr9. We expect low counts


# get the mean count for chromosome 9 on Slide 1

chr9_S1_count <- S1 %>% filter(category %in% "invariant") %>%
  filter(Size == 600) %>%
  filter(chr %in% "chr9") %>% 
  select(count) %>%
  as.matrix() %>%
  geometric.mean()
chr9_S1_count

# plot counts corrected by the mean chromosome 9 count

chr9_probes <- S1 %>%
  filter(chr %in% "chr9") %>%
  filter(Size == 600) %>%
  ggplot(aes(x = geneName, y = log2(count/chr9_S1_count), col = Tissue)) + geom_jitter(size = 2) +
  theme(axis.text.x = element_text(angle = 90)) +
 # scale_y_continuous(limits = c(0,6))+
  labs(title="chromosome 9")
chr9_probes 
#chr9_probes %>% ggsave(filename = "chromosome9_probes.png",device="png")

```


```{r fig.cap="Chromosome 7 probe counts normalized to the counts for invariant probes on that chromosome."}


# get the mean count for chromosome 7 on Slide 1

chr7_S1_count <- S1 %>% filter(category %in% "invariant") %>%
  filter(Size == 600) %>%
  filter(chr %in% "chr7") %>% 
  select(count) %>%
  as.matrix() %>%
  geometric.mean()
chr7_S1_count

# plot counts corrected by the mean chromosome 9 count

chr7_probes <- S1 %>%
  filter(chr %in% "chr7") %>%
  filter(Size == 600) %>%
  ggplot(aes(x = geneName, y = log2(count/chr7_S1_count), col = Tissue)) + geom_jitter(size = 1) +
  theme(axis.text.x = element_text(angle = 90)) +
  #scale_y_continuous(limits = c(0,6))+
  labs(title="chromosome 7")
chr7_probes

```

## ROI sizes and counts.

- For a full analysis, should divide by ERCC counts at each ROI size.

```{r}

S1 %>% filter(Tissue %in% "BT474") %>%
  filter(category %in% "variant") %>%
  ggplot(aes(x = geneName, y = log2(count), shape = as.factor(Size))) + geom_jitter(size = 2,alpha=0.5) +
  theme(axis.text.x = element_text(angle = 90)) 


# find the geometric mean of the negative counts
neg_mean_S1_50 <- geometric.mean(
  S1 %>% 
    filter(category %in% "ercc") %>% 
    filter(Size==50) %>% 
    select(count) %>% as.matrix()
  )
neg_mean_S1_50

# find the sd of the negative counts
neg_sd_S1_50 <- sd(
  S1 %>% 
    filter(category %in% "ercc") %>% 
    filter(Size==50) %>% 
    select(count) %>% as.matrix()
  )
neg_sd_S1_50

# find the geometric mean of the chromosome counts
inv_mean_S1_50 <- geometric.mean(
  S1 %>% 
    filter(category %in% "invariant") %>% 
    filter(Size==50) %>% 
    select(count) %>% as.matrix()
  )
inv_mean_S1_50

# plot 50 um ROIs

# S1 %>% filter(Tissue %in% "BT474") %>%
#   filter(category %in% "variant") %>%
#   filter(Size == 50) %>%
#   ggplot(aes(x = geneName, y = log2(count/inv_mean_S1_50))) + geom_jitter() +
#   geom_hline(yintercept = log2(neg_mean_S1_50/inv_mean_S1_50)) 

```

## Plot tiles per probe

- There are a couple outlier probes. E.g., one for ERBB2 and at least one for NKX2.1

```{r fig.cap="Variant probe counts per tile. Slide 1."}
tiles <- S1 %>% 
  filter(category %in% c("variant")) %>%
  filter(Size == 600) %>%
  ggplot(aes(x = start, y = log2(count/ercc_geomean_S1), col = Tissue)) +
  geom_jitter( size = 2 ) +
    theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  facet_wrap( ~ geneName, strip.position = "bottom", scales = "free_x") +
  theme_bw(base_size = 7) +
  labs(x = "tile", y = "log2( count/geomean negative count)")

tiles
ggsave(tiles,filename = "tiles_per_locus.png",device="png",width = 10, height = 6)
```

## Heatmaps

```{r heamapFunctions}
#' create heatmap of normalized count data
#'
#' @param dataset # data to plot
#' @param category_to_plot # as in "variant", "invariant", "ercc"
#' @export output plot

heater_norm <- function(dataset, category_to_plot){
  toPlot <- dataset %>% filter(category %in% category_to_plot ) %>%
    dplyr::filter(Size == 600) %>% select(Tissue,geneName,count) %>% 
    dplyr::group_by(Tissue,geneName) %>%
    dplyr::summarise(countz = log2(mean(count)/ercc_geomean_S1)) %>%
    pivot_wider(names_from = Tissue, values_from = countz) %>% 
    distinct() %>% 
    data.frame(row.names = "geneName")
  
  pheatmap(data.matrix(toPlot), fontsize = 8)  
}

#' create heatmap of raw count data
#'
#' @param dataset # data to plot
#' @param category_to_plot # as in "variant", "invariant", "ercc"
#' @export output plot

heater <- function(dataset, category_to_plot){
  toPlot <- dataset %>% filter(category %in% category_to_plot ) %>%
    dplyr::filter(Size == 600) %>% select(Tissue,geneName,count) %>% 
    dplyr::group_by(Tissue,geneName) %>%
    dplyr::summarise(countz = mean(count)) %>%
    pivot_wider(names_from = Tissue, values_from = countz) %>% 
    distinct() %>% 
    data.frame(row.names = "geneName")
  
  pheatmap(data.matrix(toPlot), fontsize = 8)  
}

#' create heatmap of raw count data
#'
#' @param dataset # data to plot
#' @param category_to_plot # as in "variant", "invariant", "ercc"
#' @export output plot

heater_ungrouped <- function(dataset, category_to_plot){
  toPlot <- dataset %>% filter(category %in% category_to_plot ) %>%
    dplyr::filter(Size == 600) %>% select(Tissue,Full_ID,count) %>% 
    dplyr::group_by(Tissue,Full_ID) %>%
    dplyr::summarise(countz = mean(count)) %>%
    pivot_wider(names_from = Tissue, values_from = countz) %>% 
    distinct() %>% 
    data.frame(row.names = "Full_ID")
  
  pheatmap(data.matrix(toPlot), fontsize = 8)  
}

```

```{r fig.cap="Heatmap of raw counts for slide 1 variants."}
heater(S1,"variant") %>% ggsave(filename = "variants_raw_counts_heatmap.png",device = "png")

```

```{r fig.cap="Heatmap of raw counts for slide 1 variants."}

# heater(S1 %>% 
#          filter(Tissue != "CD45L") %>%
#          filter(Tissue != "CD45H"),"variant") %>%
#   ggsave(filename="variants_raw_counts_cell_lines_s1.png",device = "png", height = 7, width = 7)

heater(S1,"variant") %>% ggsave(filename = "variants_raw_counts_heatmap.png",device = "png")

```


```{r fig.cap="Raw counts of variants for slide 1. Ungrouped each tile.", fig.height=11,fig.width=11}

# ungrouping the variant probes
heater_ungrouped(S1,"variant") %>% ggsave(filename = "variants_ungrouped_raw_counts_S1.png",
                                          device = "png", height = 11, width = 8)
```

```{r fig.cap="Raw counts of variants for slide 2."}

heater(counts_assigned %>% filter(Slide == 2),"variant") %>% 
  ggsave(filename = "variants_raw_counts_heatmap_s2.png",device = "png")

```

```{r fig.cap="Raw counts of variants after removing some outlier tiles.", fig.height=11,fig.width=11}

# Some of the variant probes have outlier counts. Remove
heater_ungrouped(S1 %>% 
                   filter(Full_ID != "ERBB2_3897.1:285") %>%
                   filter(Full_ID != "NKX2_1_3762.1:214")%>%
                   filter(Full_ID != "ZNF217_Hg19_8848.1:30"),"variant") %>%
  ggsave(filename = "variants_ungrouped_raw_counts_S1_outliersRem.png",
                                          device = "png", height = 11, width = 8)
```

```{r fig.cap="Raw counts of invariants on slide 1. Removed two probes with unusually high counts in some lines. These could be high counts due to choromosomal abnormalities, though."}
# the invariants have two probes with very high counts in a couple lines:
# remove those
heater(S1 %>% 
         filter(geneName != "gap0313_497807.1:22") %>%
         filter(geneName != "gap0104_391998.1:0"),"invariant") %>% 
  ggsave(filename = "invariants_raw_counts_heatmap.png", device = "png", height = 11, width = 8)

```

```{r fig.cap="Raw counts of invariants on slide 1. Removed SKBR3 because it has high counts across the board. Removed the two probes with high counts, too."}
# try plotting S1 without SKBR3
heater(S1 %>% 
         filter(Tissue != "SKBR3") %>%
         filter(geneName != "gap0313_497807.1:22") %>%
         filter(geneName != "gap0104_391998.1:0"),"invariant") %>% 
  ggsave(filename = "invariants_raw_counts_heatmap_S1_noSKBR3.png", device = "png", height = 11, width = 8)
```

```{r fig.cap="Raw counts of invariants. Slide 3."}
heater(counts_assigned %>% 
         filter(Slide == 3) %>%
         filter(geneName != "gap0313_497807.1:22") %>%
          filter(geneName != "gap0104_391998.1:0"),"invariant")
```

```{r fig.cap="ERCC counts for slide 1."}
heater(S1,"ercc") %>% ggsave(filename = "ercc_raw_counts_heatmap.png", device = "png", height = 11, width = 8)
```

```{r fig.cap="Raw counts of variants for slide 5."}

heater(counts_assigned %>% 
         filter(Slide == 5) %>%
  #       filter(Tissue != "MCF7") %>%
         filter(Tissue != "CD45H") %>%
         filter(Tissue != "CD45L"),"variant") %>% 
  ggsave(filename = "variants_raw_counts_heatmap_s5.png",device = "png")
```


```{r fig.cap="ERCC counts for slide 5."}
heater(counts_assigned %>% 
         filter(Slide == 5),"ercc") %>% 
  ggsave(filename = "ercc_raw_counts_heatmap_s5.png",device = "png", height = 11, width = 8)
```

```{r echo=FALSE, eval=FALSE}
# toMap <- S1 %>% filter(category %in% "invariant") %>%
#   dplyr::filter(Size == 600) %>% select(Tissue,geneName,count) %>% 
#   dplyr::group_by(Tissue,geneName) %>%
#   dplyr::summarise(countz = mean(count))
# 
# # switch to wide format
# converted <- toMap %>% pivot_wider(names_from = Tissue, values_from = countz)
# 
# #
# distinctCounts <- converted %>% distinct()
# distinctCounts
# 
# # convert to df and take row names from first col
# counts_df <- data.frame(distinctCounts, row.names = "geneName")
# counts_df %>% head()
# 
# # make the heatmap
# counts_heatmap <- pheatmap(data.matrix(counts_df),fontsize = 8)
# counts_heatmap
```

## Correlation between ERCC counts and invariant counts

- It seems like the ERCC counts are high in some samples (slides/tissues) but not others. Same goes for the invariants.
- Are the counts of the two correlated? 


```{r fig.cap="ERCC and invariant count correlation."}
grouped_means <- counts_assigned %>%
  filter(Size == 600) %>%
  group_by(AOI, Slide, category) %>%
  summarise(mean_count = mean(count))
grouped_means

# grouped_means %>%
#   filter(Slide == 1) %>%
#   filter(category != "multicopy") %>%
#   ggplot(aes(x = AOI, y = mean_count,col = category)) + 
#   geom_point()

# do a pivot, melt and spread to get the data into the right format
grouped_means_spread <- grouped_means %>% pivot_wider(names_from = c(AOI,Slide),values_from = mean_count) %>%
  reshape2::melt() %>%
  spread(key = category, value = value) 

# introduce a slide number variable
grouped_means_spread <- grouped_means_spread %>% dplyr::mutate(SlideNumber = str_extract(as.character(variable), "[^_]+$"))

# plot the mean invariant count to the mean ERCC count
inv_to_ercc <- grouped_means_spread %>%  
  ggplot(aes(x = invariant, y = ercc, col = SlideNumber)) + geom_point() +
  labs(x = "mean invariant count within AOI", y = "mean ERCC count within AOI") 
inv_to_ercc
inv_to_ercc %>% ggsave(filename = "invariants_vs_ercc.png", device = "png")

# they are highly correlated
cor.test(x = grouped_means_spread$invariant, y = grouped_means_spread$ercc)
```

## How does SNR differ at each concentration?

```{r}

#SNR = target counts / geomean (neg ctl counts)

# let's look at it gene-by-gene
  SNR_gene <- function(dataset, gene_name){
  
  # calculate the ERCC mean
  ## find the geometric mean of the negative counts
    ercc_geomean <- geometric.mean(
      dataset %>% 
        filter(category %in% "ercc") %>% 
        filter(Size==600) %>% 
        select(count) %>% as.matrix()
      )
    ercc_sd <- sd(
      dataset %>% 
        filter(category %in% "ercc") %>% 
        filter(Size==600) %>% 
        select(count) %>% as.matrix()
      )
 
     gene_count <- dataset %>% 
        filter(category %in% "variant") %>%
        filter(geneName %in% gene_name) %>%
        filter(Size==600) %>% 
        select(count) %>% as.matrix()

  tib <- data.frame(gene_count/ercc_sd)
   return(tib)
}
```

```{r fig.cap="SNR for NKX2-1. Slide 1."}
# make a plot for NKX2_1
# dimensions are 140 because 14 ROIs > 600 um * 10 probes per gene

snr_S1 <- SNR_gene(S1, "NKX2_1") %>% dplyr::rename(conc_8nM = count)
snr_S2 <- SNR_gene(S2, "NKX2_1") %>% dplyr::rename(conc_16nM = count)
snr_S3 <- SNR_gene(S3, "NKX2_1") %>% dplyr::rename(conc_16nM92C = count)
snr_S5 <- SNR_gene(S5, "NKX2_1") %>% dplyr::rename(conc_32nM = count)

nkx_snr <- bind_cols(snr_S1,snr_S2,snr_S3,snr_S5)
nkx_snr_melt <- nkx_snr %>% melt()

ggplot(nkx_snr_melt, aes(x = value,col=variable)) + 
  geom_density() +
  labs(x = "SNR", title = "NKX2-1")
```

```{r fig.cap="SNR for ERBB2. Slide 1."}
# make a plot for ERBB2
# dimensions are 140 because 14 ROIs > 600 um * 10 probes per gene

erbb_S1 <- SNR_gene(S1, "ERBB2") %>% dplyr::rename(conc_8nM = count)
erbb_S2 <- SNR_gene(S2, "ERBB2") %>% dplyr::rename(conc_16nM = count)
erbb_S3 <- SNR_gene(S3, "ERBB2") %>% dplyr::rename(conc_16nM92C = count)
erbb_S5 <- SNR_gene(S5, "ERBB2") %>% dplyr::rename(conc_32nM = count)

erbb2_snr <- bind_cols(erbb_S1,erbb_S2,erbb_S3,erbb_S5)
erbb2_snr_melt <- erbb2_snr %>% reshape2::melt()

ggplot(erbb2_snr_melt, aes(x = value,col=variable)) + 
  geom_density(alpha=0.5) +
  labs(x = "SNR", title = "ERBB2")

```

```{r fig.cap="SNR for CDKN2A. Slide 1."}
# make a plot for ERBB2
# dimensions are 140 because 14 ROIs > 600 um * 10 probes per gene

CDKN2A_S1 <- SNR_gene(S1, "CDKN2A") %>% dplyr::rename(conc_8nM = count)
CDKN2A_S2 <- SNR_gene(S2, "CDKN2A") %>% dplyr::rename(conc_16nM = count)
CDKN2A_S3 <- SNR_gene(S3, "CDKN2A") %>% dplyr::rename(conc_16nM92C = count)
CDKN2A_S5 <- SNR_gene(S5, "CDKN2A") %>% dplyr::rename(conc_32nM = count)

CDKN2A_snr <- bind_cols(CDKN2A_S1,CDKN2A_S2,CDKN2A_S3,CDKN2A_S5)
CDKN2A_snr_melt <- CDKN2A_snr %>% reshape2::melt()

ggplot(CDKN2A_snr_melt, aes(x = value,col=variable)) + 
  geom_density(alpha=0.5) +
  labs(x = "SNR", title = "CDKN2A")

```

- Slide 3 (92C) has higher counts than Slide 2. 
  - Does it have higher SNR?



## How do invariants on the same chromosome behave?


```{r fig.width = 11, fig.height=11}

inv_plot <- function(dataset, category_to_plot, tissue_choice){
  
  # calculate the ERCC mean and SD
  ## find the geometric mean of the negative counts
    ercc_geomean <- geometric.mean(
      dataset %>% 
        filter(category %in% "ercc") %>% 
        filter(Size==600) %>% 
        select(count) %>% as.matrix()
      )
  ## find the sd
    ercc_sd <- sd(
      dataset %>% 
        filter(category %in% "ercc") %>% 
        filter(Size==600) %>% 
        select(count) %>% as.matrix()
      )
  
  # count colors
    colourCount = length(unique(dataset$chr))
    #getPalette = colorRampPalette(brewer.pal(9, "Spectral"))
    pal = qualpal(colourCount, colorspace=list(h=c(0,360), s=c(0.3,1), l=c(0.2,0.8)))
    
  # generate the plot
  dataset %>% 
    filter(category %in% category_to_plot) %>% 
    filter(Tissue %in% tissue_choice) %>%
    filter(Size == 600) %>% 
    ggplot(aes(x = geneName, y = log2(count), col = chr)) + geom_point(size = 1) +
      theme(axis.text.x = element_text(angle = 90), axis.text=element_text(size=7)) +
      labs(title=paste(category_to_plot,tissue_choice,sep = " - Tissue: ")) +
      geom_hline(yintercept=log2(ercc_geomean)) +
      geom_hline(yintercept = log2(ercc_geomean + 2*ercc_sd), color = "blue",lty = 2) +
      scale_color_manual(values=pal$hex)   
    #scale_fill_manual(values = colorRampPalette(brewer.pal(12,"Paired"))(25))
    #scale_colour_viridis_d(option = "plasma","Chromosome")
}
```

```{r fig.cap="Invariant counts for slide 1 and two cell lines."}
cell_chr_1 <- ggarrange(
  inv_plot(S1,"invariant","MDAMB468") + theme(axis.text.x = element_blank(),legend.position = "none" ), 
  inv_plot(S1,"invariant","BT474")) 
ggsave(cell_chr_1, filename = "cell_lines_invariants_slide1.png", device = "png", width = 8, height = 8)

```

```{r fig.cap="Invariant counts for slide 5 and two cell lines."}
cell_chr_5 <- ggarrange(
  inv_plot(S5,"invariant","MDAMB468") + theme(axis.text.x = element_blank(),legend.position = "none" ), 
  inv_plot(S5,"invariant","BT474") ) 
ggsave(cell_chr_5, filename = "cell_lines_invariants_slide5.png", device = "png", width = 8, height = 8)

```

```{r fig.cap="Invariant counts for breast cancer tissue on slide 1"}
tiss_chr <- ggarrange(
  inv_plot(S1,"invariant","CD45H") + theme(axis.text.x = element_blank(),legend.position = "none" ), 
  inv_plot(S1,"invariant","CD45L") ) 


ggsave(tiss_chr,filename = "tissues_invariants.png", device = "png", width = 8, height = 8)
```

### Invariant observations:

- Chromosome Y counts should be low, but they're generally not lower than other chromosomes. 


```{r}

counts_assigned %>% filter(chr %in% c("chrX","chrY")) %>%
  remove_missing() %>%
  ggplot(aes(y = count, x=as.factor(Slide),fill=chr)) + geom_boxplot() +
  labs(x = "Slide", y = "Count") +
  ggtitle("X vs Y chromosome counts in all tissues")



counts_assigned %>% filter(chr %in% c("chrX","chrY")) %>%
  filter(Tissue %in% c("CD45H","CD45L")) %>%
  remove_missing() %>%
  ggplot(aes(y = count, x=as.factor(Slide),fill=chr)) + geom_boxplot() +
  labs(x = "Slide", y = "Count") +
  ggtitle("X vs Y chromosome counts in breast cancer tissue")


```


## Cell counts

For each cell line in slides 1-3, I counted cells in a FOV using an ImageJ macro

```
setAutoThreshold("Default dark");
//run("Threshold...");
setOption("BlackBackground", false);
run("Convert to Mask", "method=Triangle background=Dark calculate");
run("Analyze Particles...", "size=250-Infinity circularity=0.20-1.00 display stack");
```

- In some cases, I had to manually adjust the threshold:
  - `Image > Adjust > Threshold`
  - `Analyze > Analyze particles`
	  - Set threshold 250-Infinity (pix^2)

-If cells are running together, you can apply the threshold (converts to binary), then `Process > Binary > Watershed`, which will split apart cells at boundaries


```{r}
cell_count_table <- counts_assigned %>% select(Slide,Size,Tissue,Cell_Count) %>% remove_missing() %>% 
  group_by(Slide,Tissue,Size) %>%
  summarise(CellCount = mean(Cell_Count))

cell_count_table %>% 
  filter(Size == 600) %>%
  filter(Tissue != "CD45H") %>%
  ggplot(aes(x = as.factor(Slide), y = CellCount, col = Tissue)) + 
    geom_point(size = 3, alpha = 0.6) +
    labs(x = "Slide", y = "Cell Count")
```




## Normalization

- Cell line normalization:

```{r}
cellLines <- unique(counts_assigned %>% select(Tissue)) %>% filter(Tissue != "NTC")


cellLine_geomean <- function(line,slide){
  counts_assigned %>% 
  filter(Size == 600) %>%
  filter(Slide == slide) %>%
  filter(Tissue %in% line) %>%
  filter(category %in% "invariant") %>%
  select(count) %>%
  as.matrix() %>%
  geometric.mean()
}

#cellLine_geomean("SKBR3",1)


gm_1<-do.call(rbind.data.frame,
  lapply(c("SKBR3","BT474","MCF7","MDAMB468", "MDAMB231", "CD45H", "CD45L"),cellLine_geomean,1))

gm_1 <- gm_1 %>% dplyr::rename(invariant_geomean = names(gm_1)[[1]])

gm_cb_1 <- bind_cols(cellLines,gm_1)
gm_cb_1


S1_corr <- S1 %>% left_join(gm_cb_1,by = "Tissue") %>% mutate(corr_count = count/invariant_geomean)


var_corr <- S1_corr %>% 
    filter(category %in% "variant") %>% 
    filter(Size == 600) %>% 
    ggplot(aes(x = geneName, y = corr_count * 2, col = Tissue)) + geom_boxplot() +
    theme(axis.text.x = element_text(angle = 90)) +
    scale_y_continuous(limits = c(0,14)) +
    labs(title="Variants corrected for invariant geomean", 
         y = "",
         x = "Gene")

inv_corr <- S1_corr %>% 
    filter(category %in% "invariant") %>% 
    filter(Size == 600) %>% 
    ggplot(aes(x = chr, y = corr_count * 2)) + geom_boxplot() +
    theme(axis.text.x = element_text(angle = 90)) +
    scale_y_continuous(limits = c(0,14)) +
    labs(title="Invariants",y="")

ercc_corr <- S1_corr %>% 
    filter(category %in% "ercc") %>% 
    filter(Size == 600) %>% 
    ggplot(aes(x = geneName, y = corr_count * 2)) + geom_boxplot() +
    theme(axis.text.x = element_blank()) +
    scale_y_continuous(limits = c(0,14)) +
    labs(title="ERCC", y ="")

corr_count_fig_S1 <- ggarrange(var_corr,inv_corr,ercc_corr, 
                               ncol = 3, align="h", 
                               common.legend=TRUE,
                                widths = c(2, 1, 1)) 

corr_count_fig_S1
  
annotate_figure(corr_count_fig_S1,
  left = text_grob("2*Counts in tissue / geomean(invariants in tissue)",rot = 90)) %>% 
  ggsave(filename = "corrected_counts_slide1_wide.png", device = "png", height = 7, width =14)

```

## Cell count vs. invariant count

```{r}

S1_cell_count <- S1_corr %>% 
  filter(Size ==600) %>% 
  select(Tissue,Cell_Count) %>% remove_missing() %>% 
  group_by(Tissue) %>%
  summarise(CellCount = mean(Cell_Count))

S1_cells_to_invar <- left_join(S1_cell_count, gm_cb_1, by="Tissue") 
S1_cells_to_invar %>% ggplot(aes(x = CellCount, y = invariant_geomean)) + geom_point()


```



